File: java/src/main/java/QuizzyController.java
import java.util.*;

public class QuizzyController {
    private Collection<String> users;
    private HashMap<UUID, String> openQuizzes;
    private HashMap<UUID, String> startedQuizzes;
    private HashMap<String, Quiz> allQuizzes;
    private Set<UUID> finishedQuizzes;
    private HashMap<UUID, Collection<String>> joinedPlayers;
    private HashMap<UUID, HashMap<Integer, Collection<Tuple<String, Tuple<String, Integer>>>>> answers;
    private HashMap<UUID, Integer> currentQuestion;

    public QuizzyController() {
        users = new ArrayList<>();
        users.add("Mathias");
        users.add("Eric");
        users.add("Alberto");
        users.add("Martin");

        openQuizzes = new HashMap<>();

        SortedMap<String, String> questions = new TreeMap<>();
        questions.put("question1", "answer1");
        questions.put("question2", "answer2");

        allQuizzes = new HashMap<>();
        allQuizzes.put("star wars", new Quiz("star wars", questions));
        startedQuizzes = new HashMap<>();
        joinedPlayers = new HashMap<>();
        finishedQuizzes = new HashSet<UUID>();
        answers = new HashMap<>();
        currentQuestion = new HashMap<>();
    }

    public UUID open(String name)
    {
        if (allQuizzes.containsKey(name))
        {
            UUID gameId = UUID.randomUUID();
            openQuizzes.put(gameId, name);
            joinedPlayers.put(gameId, new ArrayList<>());
            return gameId;
        }
        else
        {
            throw new IllegalArgumentException("Quiz does not exist");
        }
    }

    public Collection<String> games()
    {
        return openQuizzes.values();
    }

    public void startGame(UUID quizId)
    {
        if (!openQuizzes.containsKey(quizId))
        {
            throw new IllegalArgumentException("Game does not exist");
        }

        if (joinedPlayers.get(quizId).size() == 0)
        {
            finishedQuizzes.add(quizId);
        }
        else
        {
            startedQuizzes.put(quizId, "");
            answers.put(quizId, new HashMap<Integer, Collection<Tuple<String,Tuple<String,Integer>>>>());
            currentQuestion.put(quizId, 0);
        }
    }

    public void join(UUID gameId, String userName)
    {
        if (!users.contains(userName))
        {
            throw new IllegalArgumentException("Player does not exist");
        }

        if (!openQuizzes.containsKey(gameId))
        {
            throw new IllegalArgumentException("Game does not exist");
        }

        if (startedQuizzes.containsKey(gameId))
        {
            throw new IllegalArgumentException("Game already started");
        }

        if (joinedPlayers.get(gameId).contains(userName))
        {
            throw new IllegalArgumentException("Player joined already");
        }

        if (finishedQuizzes.contains(gameId))
        {
            throw new IllegalArgumentException("Game is finished");
        }

        joinedPlayers.get(gameId).add(userName);
    }

    public String status(UUID quizId)
    {
        if (finishedQuizzes.contains(quizId))
        {
            return "Finished";
        }

        if (startedQuizzes.containsKey(quizId))
        {
            return "Started";
        }

        if (openQuizzes.containsKey(quizId))
        {
            return "Open";
        }

        throw new IllegalArgumentException("Game does not exist");
    }

    public String question(UUID quizId)
    {
        if (!openQuizzes.containsKey(quizId))
        {
            throw new IllegalArgumentException("Game does not exist");
        }

        if (finishedQuizzes.contains(quizId))
        {
            throw new IllegalArgumentException("Game is finished");
        }

        if (!startedQuizzes.containsKey(quizId))
        {
            throw new IllegalArgumentException("Game is not started");
        }

        Integer questionId = currentQuestion.get(quizId);
        String quizName = openQuizzes.get(quizId);
        Quiz quiz = allQuizzes.get(quizName);
        String question = (String)quiz.questions.keySet().toArray()[questionId];
        return question;
    }

    public void answer(UUID quizId, String userName, String answer, Integer answerTime)
    {
        if (!openQuizzes.containsKey(quizId))
        {
            throw new IllegalArgumentException("Game does not exist");
        }

        if (finishedQuizzes.contains(quizId))
        {
            throw new IllegalArgumentException("Game is finished");
        }

        if (!startedQuizzes.containsKey(quizId))
        {
            throw new IllegalArgumentException("Game is not started");
        }

        if (!joinedPlayers.get(quizId).contains(userName))
        {
            throw new IllegalArgumentException("Player has not joined");
        }

        HashMap<Integer, Collection<Tuple<String, Tuple<String, Integer>>>> answers = this.answers.get(quizId);
        Integer question = currentQuestion.get(quizId);
        if (answers.containsKey(question))
        {
            Collection<Tuple<String, Tuple<String, Integer>>> a = answers.get(question);

            if (a.stream().anyMatch((tuple) -> tuple.item1 == userName))
            {
                throw new IllegalArgumentException("question already answered");
            }

            a.add(new Tuple<>(userName, new Tuple<>(answer, answerTime)));
        }
        else
        {
            List<Tuple<String, Tuple<String, Integer>>> newAnswer = new ArrayList<>();
            newAnswer.add(new Tuple<>(userName, new Tuple<>(answer, answerTime)));
            answers.put(question, newAnswer);
        }

        if (answers.get(question).size() == joinedPlayers.get(quizId).size())
        {
            question = question + 1;
            currentQuestion.put(quizId, question);
        }

        String quizName = openQuizzes.get(quizId);
        Quiz quiz = allQuizzes.get(quizName);

        if (question >= quiz.questions.size())
        {
            finishedQuizzes.add(quizId);
        }
    }

    public String winner(UUID gameName)
    {
        try
        {
            if (status(gameName) == "Finished")
            {
                if (joinedPlayers.get(gameName).size() == 0)
                    throw new IllegalArgumentException("Game is finished but no players");

                String winner = "";
                Integer correctAnswerScore = 10;
                HashMap<String, Integer> scores = new HashMap<>();
                String quizName = openQuizzes.get(gameName);
                Quiz quiz = allQuizzes.get(quizName);

                for (int i = 0; i < answers.get(gameName).size(); i++) {
                    String correct = (String)quiz.questions.values().toArray()[i];
                    Collection<Tuple<String, Tuple<String, Integer>>> tuples = answers.get(gameName).get(i);

                    int finalI = i;
                    tuples.stream().forEach(tuple ->
                            {
                                String user = tuple.item1;
                                String answer = tuple.item2.item1;
                                Integer speed = tuple.item2.item2;

                    if (correct == answer)
                    {
                        Integer score = scores.getOrDefault(user, -1);
                        if (score > 0) {
                            if (speed == 0)
                            {
                                scores.put(user, score + (correctAnswerScore * (finalI +1)));
                            }
                            else
                            {
                                scores.put(user, score + ((correctAnswerScore * (finalI +1))/ speed));
                            }
                        }
                        else
                        {
                            scores.put(user, correctAnswerScore / speed);
                        }
                    }
                        });
                }

                Integer largest = 0;
                for (int i = 0; i < scores.values().size(); i++) {
                    Object user = (scores.keySet().toArray()[i]);
                    Integer score = (Integer)scores.values().toArray()[i];
                    if (score > largest)
                    {
                        largest = score;
                        winner = (String)user;
                    }
                }

                return winner;
            }

            if (status(gameName) == "Open")
            {
                throw new IllegalArgumentException("Game is not started");
            }

            if (status(gameName) == "Started")
            {
                throw new IllegalArgumentException("Game is started");
            }


            return "";
        }
        catch (IllegalArgumentException e)
        {
            throw e;
        }
    }

}

File: js/src/QuizzyController.js
const uuid4 = require("uuid/v4");

class Quiz
{
    constructor(name, questions) {
        this.name = name;
        this.questions = questions;
    }
}

class ArgumentException {
    constructor(message) {
        this.Message = message;
    }
}


class QuizzyController 
{
    constructor()
    {
        this.users = ["Mathias", "Eric", "Alberto", "Martin"];
        this._openQuizzes = {};
        var questions = [{"question1": "answer1"}, {"question2": "answer2"}];
        this._allQuizzes = {};
        this._allQuizzes["star wars"] = new Quiz("star wars", questions);
        this._startedQuizzes = {};
        this._joinedPlayers = {};
        this._finishedQuizzes = {};
        this._answers = {};
        this._currentQuestion = {};
    }

    Open(name)
    {
        if (this._allQuizzes[name])
        {
            var gameId = uuid4();
            this._openQuizzes[gameId] = name;
            this._joinedPlayers[gameId] = [];
            return gameId;
        }
        else
        {
            throw new ArgumentException("Quiz does not exist");
        }
    }

    Games()
    {
        return Object.values(this._openQuizzes);
    }

    StartGame(quizId)
    {
        if (!this._openQuizzes[quizId])
        {
            throw new ArgumentException("Game does not exist");
        }

        if (this._joinedPlayers[quizId].length === 0)
        {
            this._finishedQuizzes[quizId] = true;
        }
        else
        {
            this._startedQuizzes[quizId] = true;
            this._answers[quizId] = {};
            this._currentQuestion[quizId] = 0;
        }
    }

    Join(gameId, userName)
    {
        if (this.users.indexOf(userName) === -1)
        {
            throw new ArgumentException("Player does not exist");
        }

        if (!this._openQuizzes[gameId])
        {
            throw new ArgumentException("Game does not exist");
        }

        if (this._startedQuizzes[gameId])
        {
            throw new ArgumentException("Game already started");
        }

        if (this._joinedPlayers[gameId].indexOf(userName) !== -1)
        {
            throw new ArgumentException("Player joined already");
        }

        if (this._finishedQuizzes[gameId])
        {
            throw new ArgumentException("Game is finished");
        }


        this._joinedPlayers[gameId].push(userName);
    }

    Status(quizId) {
        if (this._finishedQuizzes[quizId])
        {
            return "Finished";
        }

        if (this._startedQuizzes[quizId])
        {
            return "Started";
        }

        if (this._openQuizzes[quizId])
        {
            return "Open";
        }

        throw new ArgumentException("Game does not exist");
    }

    Question(quizId)
    {
        if (!this._openQuizzes[quizId])
        {
            throw new ArgumentException("Game does not exist");
        }

        if (this._finishedQuizzes[quizId])
        {
            throw new ArgumentException("Game is finished");
        }

        if (!this._startedQuizzes[quizId])
        {
            throw new ArgumentException("Game is not started");
        }

        var questionId = this._currentQuestion[quizId];
        var quizName = this._openQuizzes[quizId];
        var quiz = this._allQuizzes[quizName];
        return Object.keys(quiz.questions[questionId])[0];
    }

    Answer(quizId, userName, answer, answerTime)
    {
        if (!this._openQuizzes[quizId])
        {
            throw new ArgumentException("Game does not exist");
        }

        if (this._finishedQuizzes[quizId])
        {
            throw new ArgumentException("Game is finished");
        }

        if (!this._startedQuizzes[quizId])
        {
            throw new ArgumentException("Game is not started");
        }

        if (!this._joinedPlayers[quizId].some(user => user === userName))
        {
            throw new ArgumentException("Player has not joined");
        }

        var answers = this._answers[quizId];
        var question = this._currentQuestion[quizId];
        if (answers[question])
        {
            var a = answers[question];

            if (a.some((tuple) => tuple[0] === userName))
            {
                throw new ArgumentException("Question already answered");
            }


            a.push([userName, [answer, answerTime]]);
        }
        else
        {
            answers[question] = [[userName, [answer, answerTime]]];
        }

        if (answers[question].length === this._joinedPlayers[quizId].length)
        {
            question = question + 1;
            this._currentQuestion[quizId] = question;
        }

        var quizName = this._openQuizzes[quizId];
        var quiz = this._allQuizzes[quizName];

        if (question >= quiz.questions.length)
        {
            this._finishedQuizzes[quizId] = true;
        }
    }

    Winner(gameName)
    {
        try
        {
            if (this.Status(gameName) === "Finished")
            {
                if (this._joinedPlayers[gameName].length === 0)
                    throw new ArgumentException("Game is finished but no players");

                var winner = "";
                var correctAnswerScore = 10;
                var scores = {};
                var quizName = this._openQuizzes[gameName];
                var quiz = this._allQuizzes[quizName];
                for (let i of Object.keys(this._answers[gameName]))
                {
                    var correct = Object.values(quiz.questions[i])[0];
                    var tuples = this._answers[gameName][i];

                    tuples.forEach(tuple =>
                    {
                        var user = tuple[0];
                        var answer = tuple[1][0];
                        var speed = tuple[1][1];

                        if (correct === answer)
                        {
                            let score = scores[user];
                            if (score)
                            {
                                if (speed === 0)
                                {
                                    scores[user] = score + (correctAnswerScore * (i +1));
                                }
                                else
                                {
                                    scores[user] = score + ((correctAnswerScore * (i+1))/ speed);
                                }
                            }
                            else
                            {
                                scores[user] = correctAnswerScore / speed;
                            }
                        }
                    });
                }

                var largest = 0;
                for (let user in scores)
                {
                    if (scores[user] > largest)
                    {
                        largest = scores[user];
                        winner = user
                    }
                }

                return winner;
            }

            if (this.Status(gameName) === "Open")
            {
                throw new ArgumentException("Game is not started");
            }

            if (this.Status(gameName) === "Started")
            {
                throw new ArgumentException("Game is started");
            }

            return "";
        } catch (e) {
            throw e;
        }
    }
}

module.exports.Quiz = Quiz;
module.exports.QuizzyController = QuizzyController;